.github/workflows/ci.yml

```diff
@@ -881,8 +881,9 @@ jobs:
           path: old-airflow
       - name: "Prepare airflow package: wheel"
         run: |
-          pip install pip==23.3.2 wheel==0.36.2 gitpython==3.1.40
-          python setup.py egg_info --tag-build ".dev0" bdist_wheel -d ../dist
+          pip install pip==23.3.2 hatch==1.9.1
```

Comment:
Do we still need to fix the pip version?

Reply:
We had experience with `pip` breaking things and I generally prefer to upgrade it manually - especially when it comes to eager upgrade and dependency resolution -  which is crucial part of the build 

It's 2/quarter thiing and we usually do it few days after it's released - though yeah over last 1.5 year or so it was a painless upgrade every time. 

The one "good" side of it is also full reproduicibility - we want to be able to fully reproduce installation of airflow - including the past images. Which means that all our builds shoudl be reproducible even 5 years from now.

This is why I am also going to still iterate on that PR and add reproducibility (including fixing hatchling version and any other build dependencies I think) - but I would need to automate that part because this one I want to be fixed at release time but automatically upgradeable at build time. Maybe that will be a good time to introduce it also for `pip` and any other `build-needed` dependencies.

Reply:
Pip does explictly introduce breaking changes, e.g. when https://github.com/pypa/pip/pull/12300 very old versions of `apache-airflow` will not be installable. So I agree, applications should generally pin their Pip version.

Reply:
> Pip does explictly introduce breaking changes, e.g. when https://github.com/pypa/pip/pull/12300 lands very old versions of airflow will not be installable from the latest Pip. So I agree, applications should generally pin their Pip version.

Indeed. That's also another reason why I am doing that change now because I know the longer we wait, the more likely it is for us to get it broken (especially that we abused setuptools and installation behaviour big time). 

And the calver-like versionig of `pip` indeed is clear about being able to break things (that's what we already saw.... So I want to get Airflow to be as close to PEP-standards as possible finally to avoid it :).

Also one other thing - I am personally following very closely what happens in `pip` repo - as we rely a lot of some of the features and miss a few, so I also use every release of `pip` as an oppotunity to review what's new and what's changed to know how it impacts us.

Simply `pip` is one of those dependencies for us that we want to **really** rely on.